modules/cli/src/main/java/org/apache/ignite/internal/cli/core/repl/completer/DynamicCompleterRegistry.java

```diff
@@ -38,60 +40,97 @@ public class DynamicCompleterRegistry {
+                if (!words[i].equals(prefixWords[i])) {
+                    return false;
+                }
+            }
+            return true;
         }
-        return true;
-    }
-
-    private boolean notSamePostfix(String[] words, String[] stopPostfixWords) {
-        return words.length > 0 && !Set.of(stopPostfixWords).contains(findLastNotEmptyWord(words));
-    }
 
-    private static class CompletionStrategy {
-        private final Predicate<String[]> predicate;
-        private final DynamicCompleter completer;
+        boolean canBeApplied(String[] words) {
+            // empty command means can be applied to all words
+            if (!conf.commandSpecific()) {
+                return canBeAppliedCommandMatch(words);
+            }
 
-        private CompletionStrategy(Predicate<String[]> predicate, DynamicCompleter completer) {
-            this.predicate = predicate;
-            this.completer = completer;
+            Optional<String[]> commandsMatch = conf.commands().stream().filter(command -> samePrefix(words, command)).findFirst();
+            return commandsMatch.isPresent() && canBeAppliedCommandMatch(words);
         }
 
-        boolean canBeApplied(String[] words) {
-            return predicate.test(words);
+        private boolean canBeAppliedCommandMatch(String[] words) {
+            String cursorWord = words[words.length - 1];
+            String lastNotEmptyWord = findLastNotEmptyWord(words);
+            String preLastNotEmptyWord = findLastNotEmptyWordBeforeWordFromEnd(words, lastNotEmptyWord);
+
+            if (cursorWord.equals(lastNotEmptyWord)) {
+                if (exclusiveDisableOptions.contains(lastNotEmptyWord) || exclusiveDisableOptions.contains(preLastNotEmptyWord)) {
+                    return false;
+                }
+            } else if (exclusiveDisableOptions.contains(lastNotEmptyWord)) {
+                return false;
+            }
+
+            if (conf.hasEnableOptions()) {
+                if (cursorWord.equals(lastNotEmptyWord)) {
+                    return conf.enableOptions().contains(lastNotEmptyWord) // command subcommand --enable-option
+                            || conf.enableOptions().contains(preLastNotEmptyWord); // command subcommand --enable-option lastWord
+                } else {
+                    return conf.enableOptions().contains(lastNotEmptyWord); // command subcommand --enable-option <space>
+                }
+            }
+
+            return !conf.hasDisableOptions()
+
+                    || (conf.hasDisableOptions()
+                    && !cursorWord.equals(lastNotEmptyWord)
+                    && !conf.disableOptions().contains(lastNotEmptyWord))
+
+                    || (conf.hasDisableOptions()
+                    && cursorWord.equals(lastNotEmptyWord)
+                    && !conf.disableOptions().contains(lastNotEmptyWord)
+                    && !conf.disableOptions().contains(preLastNotEmptyWord));
```

Comment:
This expression has the same logic as the if statements above, so maybe we can replace one with the other so they are consistent?

Reply:
This expression is about `disable options` and the above is about `enable options` and the logic statement is the opposite, actually. I will rewrite this expression with if-else chain to make them look similar and it will be easier to catch the difference.